helpers/deployer/main.go

```diff
@@ -0,0 +1,145 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"os"
+	"strings"
+	gotest "testing"
+	"time"
+
+	"github.com/mitchellh/go-testing-interface"
+
+	"github.com/terraform-google-modules/terraform-example-foundation/helpers/deployer/gcp"
+	"github.com/terraform-google-modules/terraform-example-foundation/helpers/deployer/steps"
+	"github.com/terraform-google-modules/terraform-example-foundation/helpers/deployer/stages"
+)
+
+type Cfg struct {
+	tfvarsFile    string
+	stepsFile     string
+	resetStep     string
+	quiet         bool
+	help          bool
+	listSteps     bool
+	disablePrompt bool
+	validate      bool
+	destroy       bool
+}
+
+func validatorApis() []string {
+	return []string{
+		"securitycenter.googleapis.com",
+		"accesscontextmanager.googleapis.com",
+	}
+}
+
+func parseFlags() Cfg {
+	var c Cfg
+
+	flag.StringVar(&c.tfvarsFile, "tfvars_file", "", "Full path to the Terraform .tfvars `file` with the configuration to be used.")
+	flag.StringVar(&c.stepsFile, "steps_file", ".steps.json", "Path to the steps `file` to be used to save progress.")
+	flag.StringVar(&c.resetStep, "reset_step", "", "Name of a `step` to be reset.")
+	flag.BoolVar(&c.quiet, "quiet", false, "If true, additional output is suppressed.")
+	flag.BoolVar(&c.help, "help", false, "Prints this help text and exits.")
+	flag.BoolVar(&c.listSteps, "list_steps", false, "List the existing steps.")
+	flag.BoolVar(&c.disablePrompt, "disable_prompt", false, "Disable interactive prompt.")
+	flag.BoolVar(&c.validate, "validate", false, "Validate tfvars file inputs.")
+	flag.BoolVar(&c.destroy, "destroy", false, "Destroy the deployment.")
+
+	flag.Parse()
+	return c
+}
+
+func main() {
+
+	cfg := parseFlags()
+	if cfg.help {
+		fmt.Println("Deploys the Terraform Example Foundation")
+		flag.PrintDefaults()
+		return
+	}
+
+	// load tfvars
+	globalTfvars, err := stages.ReadGlobalTfvars(cfg.tfvarsFile)
+	if err != nil {
+		fmt.Printf("# Failed to read GlobalTfvars file. Error: %s\n", err.Error())
+		os.Exit(1)
+	}
+
+	// validate Directories
+	err = stages.ValidateDirectories(globalTfvars)
+	if err != nil {
+		fmt.Printf("# Failed validating directories. Error: %s\n", err.Error())
+		os.Exit(1)
+	}
+
+	// init infra
+	gotest.Init()
+	t := &testing.RuntimeT{}
+
+	// only enable serivices if they are not already enabled
+	if globalTfvars.HasValidatorProj() {
+		var apis []string
+		gcpConf := gcp.NewGCP()
+		for _, a := range validatorApis() {
+			if !gcpConf.ApiIsEnabled(t, *globalTfvars.ValidatorProjectId, a) {
```

Comment:
Any downsides to just enabling any way since reenabling is a no-op?

Reply:
if we don't check if the APIs are already enabled, every time we run the helper with `-validate` we need to wait for the propagation of the activation of the APIs:

```bash
$ ./deployer -tfvars_file="./global.tfvars" -validate
# Enabling APIs: securitycenter.googleapis.com, accesscontextmanager.googleapis.com in validator project 'redacted'
# waiting for API propagation
# waiting for API propagation
# waiting for API propagation
# waiting for API propagation
# waiting for API propagation
# waiting for API propagation
# waiting for API propagation
# waiting for API propagation
# waiting for API propagation
# waiting for API propagation
# waiting for API propagation
# waiting for API propagation
# waiting for API propagation
# waiting for API propagation
# waiting for API propagation
# waiting for API propagation
# waiting for API propagation
# waiting for API propagation
# waiting for API propagation
# waiting for API propagation
# waiting for API propagation

# Validating tfvar file.
# Replace value for input 'org_id'
# Replace value 'example.com' for input 'domains_to_allow'
# Replace value 'example.com' for input 'essential_contacts_domains_to_allow'
# Replace value for input 'perimeter_additional_members': 'user:YOUR-USER-EMAIL@example.com'
# To use the feature to destroy the deployment created by this helper,
# please set the following flags to 'true' in the tfvars file:
# bucket_force_destroy, audit_logs_table_delete_contents_on_destroy, log_export_storage_force_destroy
```

If we check if the apis are enabled we don't need to wait

```bash
$ ./deployer -tfvars_file="./global.tfvars" -validate

# Validating tfvar file.
# Replace value for input 'org_id'
# Replace value 'example.com' for input 'domains_to_allow'
# Replace value 'example.com' for input 'essential_contacts_domains_to_allow'
# Replace value for input 'perimeter_additional_members': 'user:YOUR-USER-EMAIL@example.com'
# To use the feature to destroy the deployment created by this helper,
# please set the following flags to 'true' in the tfvars file:
# bucket_force_destroy, audit_logs_table_delete_contents_on_destroy, log_export_storage_force_destroy
```